Brokerages/Backtesting/BasicOptionAssignmentSimulation.cs

```diff
@@ -0,0 +1,196 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Orders;
+using QuantConnect.Securities;
+using QuantConnect.Securities.Option;
+using QuantConnect.Util;
+
+namespace QuantConnect.Brokerages.Backtesting
+{
+    /// <summary>
+    /// This market conditions simulator emulates exercising of short option positions in the portfolio.
+    /// Simulator implements basic no-arb argument: when time value of the option contract is close to zero 
+    /// it assigns short legs getting profit close to expiration dates in deep ITM positions. User algorithm then receives 
+    /// assignment event from LEAN. Simulator randomly scans for arbitrage opportunities every two hours or so.
+    /// </summary>
+    public class BasicOptionAssignmentSimulation : IBacktestingMarketSimulation
+    {
+        // we start simulating assignments 4 days prior to expiration
+        private readonly TimeSpan _priorExpiration = new TimeSpan(4,0,0,0);
+
+        // we focus only on deep ITM calls and puts (at least 5% away from price)
+        private const decimal _deepITM = 0.05m;
+
+        // we rescan portfolio for new contracts and expirations every month
+        private readonly TimeSpan _securitiesRescanPeriod = new TimeSpan(30, 0, 0, 0);
+
+        // we try to generate new assignments every 2 hours
+        private readonly TimeSpan _assignmentScanPeriod = new TimeSpan(0, 2, 0, 0);
+
+        // last update time
+        private DateTime _lastUpdate = DateTime.MinValue;
+        private Queue<DateTime> _assignmentScans;
+        private static Random _rand = new Random((int)12345);
+
+        /// <summary>
+        /// We generate a list of time points when we would like to run our simulation. we then return true if the time is in the list. 
+        /// </summary>
+        /// <returns></returns>
+        public bool IsReadyToSimulate(IAlgorithm algorithm)
+        {
+            if (_lastUpdate == DateTime.MinValue ||
+                algorithm.UtcTime - _lastUpdate > _securitiesRescanPeriod)
+            {
+                var expirations = algorithm.Securities.Keys
+                            .Where(x => x.ID.SecurityType == SecurityType.Option &&
+                                        x.ID.Date > algorithm.UtcTime &&
+                                        x.ID.Date - algorithm.UtcTime <= _securitiesRescanPeriod)
```

Comment:
Agreed: not precise code. Not doesn't affect the result, but I fixed it to have it clear.